source/common/tcp_proxy/upstream.cc

```diff
@@ -248,5 +239,71 @@ void HttpConnPool::onPoolReady(Http::RequestEncoder& request_encoder,
                                  info.downstreamSslConnection());
 }
 
+Http2Upstream::Http2Upstream(Tcp::ConnectionPool::UpstreamCallbacks& callbacks,
+                             const std::string& hostname)
+    : HttpUpstream(callbacks, hostname), response_decoder_(*this) {}
+
+bool Http2Upstream::isValidResponse(const Http::ResponseHeaderMap& headers) {
+  if (Http::Utility::getResponseStatus(headers) != 200) {
+    return false;
+  }
+  return true;
+}
+
+void Http2Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder, bool is_ssl) {
+  request_encoder_ = &request_encoder;
+  request_encoder_->getStream().addCallbacks(*this);
+  ASSERT(request_encoder_->http1StreamEncoderOptions() == absl::nullopt);
+
+  const std::string& scheme =
+      is_ssl ? Http::Headers::get().SchemeValues.Https : Http::Headers::get().SchemeValues.Http;
+  auto headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>(
+      {{Http::Headers::get().Method, "CONNECT"},
+       {Http::Headers::get().Protocol, Http::Headers::get().ProtocolValues.Bytestream},
+       {Http::Headers::get().Scheme, scheme},
+       {Http::Headers::get().Path, "/"},
+       {Http::Headers::get().Host, hostname_}});
+  request_encoder_->encodeHeaders(*headers, false);
+}
+
+Http1Upstream::Http1Upstream(Tcp::ConnectionPool::UpstreamCallbacks& callbacks,
+                             const std::string& hostname)
+    : HttpUpstream(callbacks, hostname), response_decoder_(*this) {}
+
+void Http1Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder,
+                                      bool /* isSSL unused here */) {
+  request_encoder_ = &request_encoder;
+  request_encoder_->getStream().addCallbacks(*this);
+
+  ASSERT(request_encoder_->http1StreamEncoderOptions() != absl::nullopt);
+  request_encoder_->http1StreamEncoderOptions().value().get().enableHalfClose();
+  auto headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>({
+      {Http::Headers::get().Method, "CONNECT"},
+      {Http::Headers::get().Host, hostname_},
+  });
+  request_encoder_->encodeHeaders(*headers, false);
+}
+
+bool Http1Upstream::isValidResponse(const Http::ResponseHeaderMap& headers) {
+  // According to RFC7231 any 2xx response indicates that the connection is
+  // established.
+  // Any 'Content-Length' or 'Transfer-Encoding' header fields MUST be ignored.
+  // https://tools.ietf.org/html/rfc7231#section-4.3.6
```

Comment:
Also super bonus points for RFC links
I was about to comment that they're not allowed but you're right - they're not allowed but should be ignored :-)

Let's make sure we have an integration test for each just to make sure for example that the headers say "chunked" and the connect payload looks like http body, that it is passed through as CONNECT payload and not dechuked by HTTP parser.

Reply:
I tried to add an integration test where the upstream server replies with `TransferEncoding` header set, but it fails on this assertion:

https://github.com/envoyproxy/envoy/blob/d0ea3325dd96510ae785d2deef25e17059540672/source/common/http/http1/codec_impl_legacy.cc#L142

Maybe I can just try to add some unit test coverage here?

Reply:
you'll probably have to use the raw connection code
see TestServerAllowChunkedLength for how to wait for request headers complete (\r\n\r\n) and send a raw HTTP/1.1 response.